### PR TITLE
Implementation of KmerFinder subworkflow Custom Quast, and Custom MultiQC Reports 

### LINKED ISSUE #109: Improvements in  forked repository: KmerFinder, Estimation of Reference Genome,  Custom Quast, and Custom MultiQC Reports
## Overview

Hello!, My colleagues and I have been actively working on enhancing the nf-core/bacass workflow to address lab-specific challenges in bacterial genome assembly. We are happy to add  these improvements into the main nf-core/bacass repository in case you are interested.

Currently, these enhancements have been implemented in my [local fork of nf-core/bacass on the buisciii-develop branch](https://github.com/Daniel-VM/bacass/tree/buisciii-develop).
```
nextflow run main.nf \
        -profile singularity,test \
        --skip_kmerfinder false \
        --kmerfinderdb path/to/kmerfinder_db/bacteria \
        --ncbi_assembly_metadata path/to/ncbi_assembly_metadata/assembly_summary_bacteria.txt \
        --outdir ./results \
        -w ./work \
        -resume
```
### 1. Kmerfinder Subworkflow:

- Added a local KmerFinder module for read quality control (QC) and purity assessment.
- Developed a local module to compile KmerFinder results from all samples into a comprehensive CSV summary file.
- Implemented a method to group input samples (*.fastq, *.fasta, and other files...) based on the reference genome estimated with KmerFinder.
- Created a local module to identify the reference genome estimated with KmerFinder in the NCBI database and download this genome. This reference genome is then utilized to retrieve relevant metrics from QUAST, such as the percentage of genome fraction. This functionality is particularly valuable when input samples belong to different species, requiring more than one reference for a comprehensive by_reference_genome report.

### 2. Quast Assembly QC by Grouping Samples:

- Modified Quast execution when KmerFinder is invoked. Now, Quast runs twice:
- Initial 'general' Quast without reference genome files (*.fna, *.gff).
- Subsequent 'by reference genome' Quast, providing a Quast reports that agregates samples and their reference genome (estimated with kmerfinder).

### 3. Custom MultiQC Reports:

- Incorporated a custom MultiQC module into the workflow.
- Added multiqc_config.yml files for short, long and hybrid assembly modes (they work when kmerfinder is invoked, otherwise standard multiqc report is generated). 
- Upon invoking KmerFinder, a custom MultiQC HTML report is generated using the MULTIQC() module. This report consolidates metrics from KmerFinder, Quast, and other relevant sources, presenting them together in an overview table located in the first section of the report. See image: 

Screenshot from 2024-01-04 16-23-26

## Foot note

If you think these improvements could be implemented in nf-core/bacass, let me know so I can work on the test data and test profile.

### PR DESCRIPTION
**Pull request by Daniel-VM**
## PR checklist

- [X] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/bacass/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/bacass _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [X] Make sure your code lints (`nf-core lint`).
- [X] Usage Documentation in `docs/usage.md` is updated.
- [X] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [X] `README.md` is updated (including new tool citations and authors/contributors).

## PR Description
This PR introduces significant enhancements to the nf-core/bacass pipeline by incorporating a KmerFinder subworkflow and improving Quast Assembly QC and MultiQC reports.

**Key Implementations**:
1. KmerFinder Subworkflow:
- Added a local KmerFinder module for read quality control (QC) and purity assessment.
- Developed a module to compile KmerFinder results into a comprehensive CSV summary.
- Implemented grouping of input samples based on the KmerFinder-estimated reference genome.
- Created a module to identify and download reference genomes from the NCBI database, used for retrieving relevant QUAST metrics.

3. Quast Assembly QC by Grouping Samples:
- Modified Quast execution to run twice when KmerFinder is invoked:
- Initial 'general' Quast without reference genome files.
- Subsequent 'by reference genome' Quast to aggregate samples and their reference genomes.

5. Custom MultiQC Reports:
- Incorporated a custom MultiQC module into the workflow.
- Added `multiqc_config.yml` files for short, long, and hybrid assembly modes.
- Generated a custom MultiQC HTML report consolidating metrics from KmerFinder, Quast, and other sources, presented in an overview table.

Closes #109 

## RUN TESTS
Test of this PR can be run via:
`nextflow run main.nf -profile test_full,docker --outdir results`

or 

```
nextflow run main.nf  \
    -profile test,docker \
    --skip_kmerfinder false \
    --kmerfinderdb 'https://zenodo.org/records/10458361/files/20190108_kmerfinder_stable_dirs.tar.gz?download=1' \
    --ncbi_assembly_metadata 'https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt' \
    --outdir results
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,6 +48,12 @@ For users specifying both short read and long read (NanoPore) data, the pipeline
 
 In all cases, the assembly is assessed using [QUAST](http://bioinf.spbau.ru/quast). The resulting bacterial assembly is furthermore annotated using [Prokka](https://github.com/tseemann/prokka), [Bakta](https://github.com/oschwengers/bakta) or [DFAST](https://github.com/nigyta/dfast_core).
 
+In specific cases where samples recorded in the input samplesheet belong to more than one species, the pipeline finds and downloads their respectve reference genomes (this also works with single specie input samplesheet). It then groups the samples into batches and collects assembly QC results based on their corresponding reference genomes.
```
Comment (d4straub) [inline README.md:51]: typos?
```suggestion
In specific cases where samples recorded in the input samplesheet belong to more than one species, the pipeline finds and downloads their respective reference genomes (this also works with single species input samplesheet). It then groups the samples into batches and collects assembly QC results based on their corresponding reference genomes.
```

File: README.md [hunk 2]
```diff
@@ -47,6 +48,12 @@ For users specifying both short read and long read (NanoPore) data, the pipeline
...
+
+> NOTE: This scenario is supported when [Kmerfinder](https://bitbucket.org/genomicepidemiology/kmerfinder/src/master/) analysis is performed only.
+
+In cases where input samplesheet has files where , the pipeline will group samples in batches according to their reference genomes and will provide a general QUAST containing all the input samples and a by reference genome QUAST report, that is, a quast report for each reference genome.
```
Comment (d4straub) [inline README.md:55]: Something seems off with the beginning of this sentence.

File: conf/modules.config [hunk 3]
```diff
@@ -31,13 +31,15 @@ process {
         ]
     }
 
+    // FIXME: SAVE TRIMMED NOT WORKING
```
Comment (d4straub) [inline conf/modules.config:34]: still something to fix?

File: conf/modules.config [hunk 4]
```diff
@@ -176,12 +178,23 @@ process {
         ]
     }
 
-    withName: 'QUAST' {
+    // FIXME: output structure and meta updated. It might require a fixme
```
Comment (d4straub) [inline conf/modules.config:181]: is that solved?

File: conf/modules.config [hunk 5]
```diff
@@ -176,12 +178,23 @@ process {
...
+    withName: 'QUAST|QUAST_BYREFSEQID' {
         ext.args = ''
         publishDir = [
             path: { "${params.outdir}/QUAST" },
             mode: params.publish_dir_mode,
-            saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
+            //pattern: "{report,runs_per_reference/*}/{report.html,report.pdf,icarus.html}",
```
Comment (d4straub) [inline conf/modules.config:187]: I guess that is solved below?

File: conf/test_full.config [hunk 6]
```diff
@@ -15,6 +15,9 @@ params {
     config_profile_description = 'Full test dataset to check pipeline function'
 
     // Input data for full size test
-    input       = params.pipelines_testdata_base_path + 'bacass/bacass_full.tsv'
-    kraken2db   = 'https://genome-idx.s3.amazonaws.com/kraken/k2_standard_8gb_20210517.tar.gz'
+    input                       = 'https://raw.githubusercontent.com/nf-core/test-datasets/bacass/bacass_full.tsv'
+    kraken2db                   = 'https://genome-idx.s3.amazonaws.com/kraken/k2_standard_8gb_20210517.tar.gz'
+    kmerfinderdb                = 'https://zenodo.org/records/10458361/files/20190108_kmerfinder_stable_dirs.tar.gz?download=1'
```
Comment (d4straub) [inline conf/test_full.config:20]: is the part `?download=1` really needed?

File: workflows/bacass.nf [hunk 7]
```diff
@@ -9,14 +9,40 @@ def checkPathParamList = [ params.input, params.multiqc_config, params.kraken2db
...
 
+// Check kmerfinder dependencies
+if (!params.skip_kmerfinder) {
+    if (!params.kmerfinderdb || !params.ncbi_assembly_metadata) {
+        exit 1, "[KMERFINDER]: Missing --kmerfinder_db and/or --ncbi_assembly_metadata arguments. Both are required to run KMERFINDER."
+    } else {
+        kmerfinderdb = file(params.kmerfinderdb, checkIfExists: true)
+        ncbi_assembly_metadata = file(params.ncbi_assembly_metadata, checkIfExists: true)
+    }
+}
```
Comment (d4straub) [inline workflows/bacass.nf:28]: I thought this validation goes usually into https://github.com/nf-core/bacass/blob/3a51224f849386e3d8aeac4c949f6ca99355a6a4/subworkflows/local/utils_nfcore_bacass_pipeline/main.nf#L158 (and https://github.com/nf-core/bacass/blob/3a51224f849386e3d8aeac4c949f6ca99355a6a4/subworkflows/local/utils_nfcore_bacass_pipeline/main.nf#L78 un-commented) but not critical imho.

File: conf/test_hybrid_dragonflye.config [hunk 8]
```diff
@@ -23,8 +23,9 @@ params {
     input = params.pipelines_testdata_base_path + 'bacass/bacass_hybrid_dragonflye.tsv'
 
     // some extra args to speed tests up
-    assembly_type   = 'hybrid'
-    assembler       = 'dragonflye'
-    prokka_args     = " --fast"
+    assembly_type   ='hybrid'
+    assembler       ='dragonflye'
+    prokka_args     =" --fast"
```
Comment (d4straub) [inline conf/test_hybrid_dragonflye.config:28]: why were the blanks removed? I'd prefer with blank space ;)

File: docs/output.md [hunk 9]
```diff
@@ -118,6 +118,22 @@ Exemplary Kraken2 report screenshot:
...
+## Reads QC and Sample purity
+
+The pipeline includes a dedicated step for short and long reads QC as well as contamination analysis using [Kmerfinder](https://bitbucket.org/genomicepidemiology/kmerfinder/src/master/). This process helps assess the quality and purity of the samples.
+
+<details markdown="1">
+<summary>Output files</summary>
+
+- `Kmerfinder/{ID}/`
+
+  - `*_results.txt`: Kmerfinder report table containing reads QC results and taxonomic information.
```
Comment (d4straub) [inline docs/output.md:130]:
```suggestion
- `Kmerfinder/{ID}/`
  - `*_results.txt`: Kmerfinder report table containing reads QC results and taxonomic information.
```

File: docs/output.md [hunk 10]
```diff
@@ -118,6 +118,22 @@ Exemplary Kraken2 report screenshot:
...
+
+- `Kmerfinder/`:
```
Comment (d4straub) [inline docs/output.md:132]:
```suggestion
- `Kmerfinder/`
```

File: docs/output.md [hunk 11]
```diff
@@ -118,6 +118,22 @@ Exemplary Kraken2 report screenshot:
...
+  - kmerfinder_summary.csv: A CSV file containing the most relevant results of all samples analyzed with Kmerfinder.
```
Comment (d4straub) [inline docs/output.md:133]:
```suggestion
  - `kmerfinder_summary.csv`: A CSV file containing the most relevant results of all samples analyzed with Kmerfinder.
```

File: docs/output.md [hunk 12]
```diff
@@ -180,9 +196,13 @@ The assembly QC is performed with [QUAST](http://quast.sourceforge.net/quast) fo
 <details markdown="1">
 <summary>Output files</summary>
 
-- `QUAST`
-  - `report.tsv`: QUAST's report in text format
-- `QUAST/report`
+- `QUAST/report/`
+
```
Comment (d4straub) [inline docs/output.md:200]:
```suggestion
```

File: modules/local/kmerfinder.nf [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        --db_path ${kmerfinder_db}/bacteria.ATG \\
+        -tax ${kmerfinder_db}/bacteria.name \\
+        -x
+
+    mv results.txt ${prefix}_results.txt
+    mv data.json ${prefix}_data.json
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        kmerfinder: \$(echo "3.0.2")
```
Comment (d4straub) [inline modules/local/kmerfinder.nf:36]: This is usually solved by defining the version number directly below script: and adding a big warning to modify when changing the container. There should be some nf-core modules as examples.

File: modules/nf-core/quast/main.nf [hunk 14]
```diff
@@ -37,7 +37,7 @@ process QUAST {
         $args \\
         ${consensus.join(' ')}
 
-    ln -s ${prefix}/report.tsv ${prefix}.tsv
+    ln -s ${prefix}/report.tsv report.tsv
```
Comment (d4straub) [inline modules/nf-core/quast/main.nf:40]: Could that be solved by setting task.ext.prefix to "report"? The advantage might be that the nf-core modules would not need any modification.

File: nextflow_schema.json [hunk 15]
```diff
@@ -72,6 +80,22 @@
...
+                    "type": "string",
+                    "description": "Reference FASTA file."
+                },
+                "reference_gff": {
+                    "type": "string",
+                    "description": "Reference GFF file."
+                },
+                "ncbi_assembly_metadata": {
+                    "type": "string",
+                    "description": "Master file (*.txt) containing a summary of asseblies available in GeneBank or RefSeq. See: https://ftp.ncbi.nlm.nih.gov/genomes/README_assembly_summary.txt"
```
Comment (d4straub) [inline nextflow_schema.json:98]:
```suggestion
                    "description": "Master file (*.txt) containing a summary of assemblies available in GeneBank or RefSeq. See: https://ftp.ncbi.nlm.nih.gov/genomes/README_assembly_summary.txt"
```

File: workflows/bacass.nf [hunk 16]
```diff
@@ -9,14 +9,40 @@ def checkPathParamList = [ params.input, params.multiqc_config, params.kraken2db
...
+/*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    CONFIG FILES
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+*/
+
+
+// When invoking kmerfinder, utilize a custom MultiQC config file to generate a specialized report. This report will organize samples into groups based on their reference genome, w were previously calculated by kmerfinder.
```
Comment (d4straub) [inline workflows/bacass.nf:36]: typo in the last part of the second sentence?

File: workflows/bacass.nf [hunk 17]
```diff
@@ -374,21 +404,79 @@ workflow BACASS {
...
+
+    ch_kmerfinder_multiqc = Channel.empty()
+    if (!params.skip_kmerfinder) {
+        // Prepare kmerfinder database
+        if ( kmerfinderdb.name.endsWith('.gz') ) {
+            UNTAR ( [[ id: kmerfinderdb.getSimpleName() ], kmerfinderdb] )
+            ch_kmerfinderdb_untar = UNTAR.out.untar.map{ meta, file -> file }
+        } else {
+            ch_kmerfinderdb_untar = Channel.from(kmerfinder_db)
+        }
```
Comment (d4straub) [inline workflows/bacass.nf:420]: Couldnt that be in the KMERFINDER_SUBWORKFLOW?

**Review by d4straub — COMMENTED** [review]
Hi there, seems really great (but unfortunately no time to properly run a test).
I have a few comments, most of them only typo.

File: README.md (hunk 2 above)
Comment (Daniel-VM) [reply README.md:55]: God, true. Updating this.

File: conf/modules.config (hunk 5 above)
Comment (Daniel-VM) [reply conf/modules.config:187]: yes!

File: conf/test_hybrid_dragonflye.config (hunk 8 above)
Comment (Daniel-VM) [reply conf/test_hybrid_dragonflye.config:28]: Not sure why I did that lol. Recovering blanks ;).

File: conf/test_full.config (hunk 6 above)
Comment (Daniel-VM) [reply conf/test_full.config:20]: Worked, thanks!!

File: workflows/bacass.nf (hunk 7 above)
Comment (Daniel-VM) [reply workflows/bacass.nf:28]: Awesome, I forgot to move it into the new nf-core templates. Much better!

File: conf/modules.config (hunk 3 above)
Comment (Daniel-VM) [reply conf/modules.config:34]: It was a typo in the "enabled" option. It is now fixed.

File: modules/nf-core/quast/main.nf (hunk 14 above)
Comment (Daniel-VM) [reply modules/nf-core/quast/main.nf:40]: I am not entirely sure. I tried running a few tests by setting `task.ext.prefix = { "report" }` in the` conf/modules.config `(QUAST section). However, this affects other steps in the pipeline such `CUSTOM_MULTIQC()`, which is not executed after setting `task.ext.prefix = { "report" }`.

Comment (Daniel-VM) [reply modules/nf-core/quast/main.nf:40]: I am going to do a more deeper review to get this enhancement implemented. 😅

Comment (Daniel-VM) [reply modules/nf-core/quast/main.nf:40]: Fixed. No patch needed anymore :).

**Comment by Daniel-VM** [comment]
Many thanks @d4straub ! Your suggestions have been very helpful. I have also added fixes because `custom_multiqc` wasn't working propertly after merging into `dev`.

**Review by d4straub — APPROVED** [review]
Great!

**Pull request merged**